metadata-ingestion/src/datahub/ingestion/source/tableau.py

```diff
@@ -1954,6 +1954,42 @@ def emit_datasource(
                     aspect_name=c.UPSTREAM_LINEAGE,
                     aspect=upstream_lineage,
                 )
+        else:
+            # Fetch upstream information from upstreamColumns which are mostly coming
+            # from CUSTOM SQLs when Tableu doesn't provide upstream tables
+            # With this approach, we also join lingering CustomSQL entities to Embedded Data Sources
+            table_id_to_urn = {
+                column[c.TABLE][c.ID]: builder.make_dataset_urn_with_platform_instance(
+                    self.platform,
+                    column[c.TABLE][c.ID],
+                    self.config.platform_instance,
+                    self.config.env,
+                )
+                for field in datasource.get(c.FIELDS, [])
+                for column in field.get(c.UPSTREAM_COLUMNS, [])
+                if column.get(c.TABLE, {}).get(c.TYPE_NAME) == c.CUSTOM_SQL_TABLE
+                and column.get(c.TABLE, {}).get(c.ID)
+            }
+            fine_grained_lineages = self.get_upstream_columns_of_fields_in_datasource(
+                datasource, datasource_urn, table_id_to_urn
+            )
+            upstream_tables = [
+                Upstream(dataset=table_urn, type=DatasetLineageType.TRANSFORMED)
+                for table_urn in table_id_to_urn.values()
+            ]
+            upstream_lineage = UpstreamLineage(
```

Comment:
let's put this new logic in the `_create_upstream_table_lineage` method

we can move this if statement into the method as well if necessary
```
        # Upstream Tables
        if (
            datasource.get(c.UPSTREAM_TABLES)
            or datasource.get(c.UPSTREAM_DATA_SOURCES)
            or datasource.get(c.FIELDS)
        ):
```

Reply:
Makes sense